File: dfs/980_unique_paths_iii.py
class Solution:
    # O(4^m*n) time | O(m*n) space
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        squares = 1
        startR, startC = 0, 0
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 0:
                    squares += 1
                elif grid[r][c] == 1:
                    startR, startC = r, c
        
        def backtrack(r, c, squares):
            if r < 0 or r == ROWS or c < 0 or c == COLS or grid[r][c] == -1:
                return 0
            if grid[r][c] == 2:
                return squares == 0
            uniquePaths = 0
            grid[r][c] = -1
            directions = [[0, 1], [0, -1], [-1, 0], [1, 0]]
            for dr, dc in directions:
                uniquePaths += backtrack(r + dr, c + dc, squares - 1)
            grid[r][c] = 0
            return uniquePaths

        return backtrack(startR, startC, squares)

File: dfs/980_unique_paths_iii.go
func uniquePathsIII(grid [][]int) int {
    ROWS, COLS := len(grid), len(grid[0])
    squares := 1
    startR, startC := 0, 0
    
    for r := 0; r < ROWS; r++ {
        for c := 0; c < COLS; c++ {
            if grid[r][c] == 0 {
                squares++
            } else if grid[r][c] == 1 {
                startR, startC = r, c
            }
        }
    }
    return backtrack(grid, startR, startC, squares)
}

func backtrack(grid [][]int, r int, c int, squares int) int {
    if r < 0 || r == len(grid) || c < 0 || c == len(grid[0]) || grid[r][c] == -1 {
        return 0
    }
    if grid[r][c] == 2 {
        if squares == 0 {
            return 1
        } else {
            return 0
        }
    }
    grid[r][c] = -1
    uniquePaths := backtrack(grid, r + 1, c, squares - 1) + backtrack(grid, r - 1, c, squares - 1) + backtrack(grid, r, c + 1, squares - 1) + backtrack(grid, r, c - 1, squares - 1)
    grid[r][c] = 0
    return uniquePaths
} 
